orc8r/cloud/go/obsidian/swagger/combine_test.go

```diff
@@ -0,0 +1,97 @@
+/*
+ Copyright 2020 The Magma Authors.
+
+ This source code is licensed under the BSD-style license found in the
+ LICENSE file in the root directory of this source tree.
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package swagger_test
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"magma/orc8r/cloud/go/obsidian/swagger"
+	"magma/orc8r/cloud/go/obsidian/swagger/protos"
+	"magma/orc8r/cloud/go/orc8r"
+	"magma/orc8r/cloud/go/test_utils"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_GetCombinedSwaggerSpecs(t *testing.T) {
```

Comment:
This test doesn't need to create any files. In fact, don't even use any golden files. Just create super-short swagger specs inline in this file. We're not testing whether combine "correctly" combines in this test, we're testing the e2e datapath is correctly hooked up

Reply:
Having inline swagger specs work as the test case was actually the first thing I considered and I actually talked this over with Michael a bit as well. The main reason why I decided to use the golden files pattern was because I found it to be actually quite "ugly" having swagger specs declared in-line, mainly just due to the fact that the minimum boilerplate tags needed (magma-gen-meta, info, base-path, etc) already takes up about 6-7 lines. Coupled with the fact that we want to at least have one unique tag per service as a general sanity test + common spec, I thought it would optimal to just stick with the golden files pattern that we have used in the swagger lib file. Note that the test files I have right now are already extremely minimal(just one unique tag each). I think the big benefit with the golden file patterns is actually in mind the simplicity? I do see how having specs inline may actually be beneficial in the sense that we want to avoid IO operations as much as possible so that we minimize any potential conflicts when running concurrent tests. Besides what I listed above, are there any main benefits that I'm missing with transitioning to in-line specs? @hcgatewood

Reply:
Update: I just realized that the IO benefit of the spec inline isn't actually there since we would have to create a temp directory to store our common-spec regardless.

Reply:
As a starting point: I/O makes tests messy. Except where we're specifically testing the I/O's correct functioning, we want to avoid including I/O in our tests where achieving this goal doesn't come with unreasonable trade-offs

In this case, I think there's a clean way to have this test do zero I/O

1. Handling the inline specs

A. The magma-gen-meta is for model generation rather than API specification, so it can be discarded from these toy specs. (In fact, a service should be able to send a generic Swagger spec and everything would work fine.) B. If you still dislike the length of the multiline strings (I'm assuming these strings will be less than 10 lines each), create the per-service specs as spec.Swagger objects first, then serialize to string before passing to the servicer

2. Handling the common file I/O

Let's rework the poll+combine command slightly to handle this elegantly. A. The `GetCombinedSpec` method can become `GetCombinedSpec(yamlCommon string) (string, error)` and B. add a method `GetCommonSpec() (string, error)`. We'll still need an I/O test for `GetCommonSpec`, but this will allow our test for `GetCombinedSpec` to be I/O-free. (It will also allow an optimization where the obsidian service can read the common spec once, during initialization, instead of re-reading it on every call.)

Notice how this rearranges things such that we're testing the I/O function in isolation, while the larger business logic function an be tested as a pure function. This is a healthy testing pattern